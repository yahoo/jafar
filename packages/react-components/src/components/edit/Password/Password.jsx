/**
  * Copyright 2020, Verizon Media
  * Licensed under the terms of the MIT license. See LICENSE file in project root for terms.
  */

import React from 'react';
import PropTypes from 'prop-types';
import Input from '@material-ui/core/Input';

/**
 * Represent a string
 * 
 * Import <a target="_blank" 
 href="https://github.com/yahoo/jafar/blob/master/packages/react-components/src/components/edit/Password/Password.jsx">
 Password</a> from '@jafar/react-components/edit/Password'
 */
export default class Password extends React.Component {
  static propTypes = {
    value: PropTypes.string,
    disabled: PropTypes.bool,
    invalid: PropTypes.bool,
    onValueChange: PropTypes.func.isRequired,
  };

  static defaultProps = {
    value: '',
    disabled: false,
    invalid: false,
  };

  render() {
    return (
      <Input
        type="password"
        value={this.props.value}
        disabled={this.props.disabled}
        error={this.props.invalid}
        onChange={this.onValueChange} />
    );
  }

  onValueChange = (e) => {
    this.props.onValueChange(e.target.value);
  }
}
