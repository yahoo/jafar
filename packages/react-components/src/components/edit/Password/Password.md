<h3>Examples</h3>

<h4>Usage in jafar form</h4>

```javascript
const Form = require('@jafar-org/react-form/Form').default;
const Field = require('@jafar-org/react-form/Field').default;

const model = {
  id: 'simple',
  fields: {
    userPassword: {
      label: 'User Password',
      path: 'u',
      component: {
        name: 'Password',
      },
      required: true,
      disableTerm: {
          name: 'disableMe',
      },
      excludeTerm: {
          name: 'excludeMe',
      },
    },
  },
  data: {
      value: 'Im@King22385', // remove to see initial value undefined, and required error
  },
};

const resources = {
  components: { 
    Password: {
      renderer: Password,
    },
  },
  terms: {
      disableMe: { func: () => false }, // convert to true to see the field disabled
      excludeMe: { func: () => false }, // convert to true to see the field excluded
  },
};

<Form model={model} resources={resources}>
    <Field id='userPassword' />
</Form>
```


<h4>Simple</h4>

```javascript
initialState = { 
    value: 'Im@King22385',
};

 <Password
    value={state.value}
    onValueChange={(value) => {
        setState({ value });
    }}
    />
```

<h4>Initial value undefined</h4>

```javascript
initialState = { 
    value: undefined,
};
 <Password
    value={state.value}
    onValueChange={(value) => {
        setState({ value });
    }}
    />
```

<h4>Disabled</h4>

```javascript
initialState = { 
    value: 'Im@King22385',
};
 <Password
    value={state.value}
    disabled={true}
    onValueChange={(value) => {
        setState({ value });
    }}
    />
```

<h4>Invalid</h4>

```javascript
initialState = { 
    value: 'Im@King22385',
};
 <Password
    value={state.value}
    invalid={true}
    onValueChange={(value) => {
        setState({ value });
    }}
    />
```