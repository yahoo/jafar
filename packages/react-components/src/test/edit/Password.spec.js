/**
  * Copyright 2020, Verizon Media
  * Licensed under the terms of the MIT license. See LICENSE file in project root for terms.
  */

import React from 'react';
import { mount, shallow } from 'enzyme';
import Password from '../../edit/Password/index';

describe('<Password />', () => {
  let component;
  let onValueChangeSpy;
  const value = 'secretPassword';
  const disabled = false;
  const invalid = false;
  const state = { };

  beforeEach(() => {
    onValueChangeSpy = jest.fn();
  });

  it('should render provided data', () => {
    component = shallow(
      getComponent(value, state, disabled, invalid, onValueChangeSpy)
    );
    expect(component).toMatchSnapshot();
  });

  it('should trigger onValueChange callback on value change', () => {
    component = mount(
      getComponent(value, state, disabled, invalid, onValueChangeSpy)
    );
    const newValue = 'newPassword';
    component.find('input').simulate('change', { target: { value: newValue } });
    expect(onValueChangeSpy).toHaveBeenCalledWith(newValue);
  });

  function getComponent(value, state, disabled, invalid, onValueChange) {
    return (
      <Password
        value={value}
        state={state}
        disabled={disabled}
        invalid={invalid}
        onValueChange={onValueChange}
      />
    );
  }
});
