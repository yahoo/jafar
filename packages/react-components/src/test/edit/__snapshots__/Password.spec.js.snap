// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Password /> should render provided data 1`] = `
<WithStyles(ForwardRef(Input))
  disabled={false}
  error={false}
  onChange={[Function]}
  type="password"
  value="secretPassword"
/>
`;
